iroh/src/magicsock/transports.rs

```diff
@@ -0,0 +1,432 @@
+use std::{
+    io::{self, IoSliceMut},
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use iroh_base::{NodeId, RelayUrl};
+use n0_watcher::Watcher;
+#[cfg(not(wasm_browser))]
+use netwatch::UdpSocket;
+use relay::RelayDatagramSendChannelSender;
+use tracing::{trace, warn};
+
+#[cfg(not(wasm_browser))]
+mod ip;
+mod relay;
+
+#[cfg(not(wasm_browser))]
+pub(crate) use self::ip::IpTransport;
+pub(crate) use self::relay::{RelayActorConfig, RelayTransport};
+use super::NetInfo;
+
+#[derive(Debug)]
+pub(crate) struct Transports {
+    #[cfg(not(wasm_browser))]
+    ip: Vec<IpTransport>,
+    relay: Vec<RelayTransport>,
+}
+
+impl Transports {
+    /// Create a new transports structure.
+    pub(crate) fn new(
+        #[cfg(not(wasm_browser))] ip: Vec<IpTransport>,
+        relay: Vec<RelayTransport>,
+    ) -> Self {
+        Self {
+            #[cfg(not(wasm_browser))]
+            ip,
+            relay,
+        }
+    }
+
+    /// Send the given [`Transmit`] to the given [`Addr`].
+    ///
+    /// Sends on the first matching & ready transport.
+    pub(crate) fn poll_send(
+        &self,
+        destination: &Addr,
+        transmit: &Transmit<'_>,
+    ) -> Poll<io::Result<()>> {
+        trace!(?destination, "sending");
+
+        match destination {
+            #[cfg(wasm_browser)]
+            Addr::Ipv4(..) => {
+                return Poll::Ready(Err(io::Error::other("IPv4 is unsupported in browser")))
+            }
+            #[cfg(not(wasm_browser))]
+            Addr::Ipv4(addr, port) => {
+                let addr = SocketAddr::V4(SocketAddrV4::new(*addr, port.unwrap_or_default()));
```

Comment:
I'm confused how you can try send to a `0` port.

Reply:
fixed